### ISSUE TITLE
Precedence issue introduced by redefinition of method in inheritor class

### ISSUE THREAD

**Issue by femtomc** [beartype/plum#130]
Hi! I'm working on a design pattern which is a bit complicated, and I'm not sure how to proceed.

![image](https://github.com/beartype/plum/assets/34410036/0c1de35b-4117-4e62-82c7-f36e235857d7)

I've attached a screenshot which shows a pattern of dispatch I'm using. I have an abstract base class which defines a "NotImplemented" dispatch base for this method `update` (top of image), then I implement a specialized version with type dispatch on the type `Mask` -- now, `Mask` inherits from `Choice`.

Later on, I use the abstract base class and implement a subclass called `SwitchCombinator`. There, I redefine the method for the type `Choice` (but not the dispatch for `Mask`)!

Now, unfortunately, this appears to clobber the dispatch -- because when I test, even if I pass in a `Mask` -- the dispatch branch is the `Choice` one.

Is there a pattern which I can use here? Does this seem like something which should be supported?

**Comment by PhilipVinc** [comment]
can you share a MWE?

**Comment by femtomc** [comment]
Sure! will prepare one and send this afternoon

**Comment by femtomc** [comment]
@PhilipVinc Here's an MWE:

```python
from plum import dispatch

class Parent:
    pass

class B(Parent):
    pass

class BaseClass:
    # Specialized to B, which inherits from Parent.
    @dispatch
    def foo(self, x: B):
        print("hello!")
        
    # Generic -- not defined yet!
    @dispatch
    def foo(self, x: Parent):
        raise NotImplementedError
        
class Inheritor(BaseClass):
    # Inheritor of the BaseClass, okay I define a fallback
    # on the generic Parent base class.
    @dispatch
    def foo(self, x: Parent):
        print("cool!")
        
# Yay, specialized!
BaseClass().foo(B())

# Oops, clobbered.
Inheritor().foo(B())
```

**Comment by wesselb** [comment]
Hey @femtomc! It's not entirely clear to me which behaviour you would want. Should `Inheritor().foo(B())` print something other than `cool!`?

It might be helpful to consider the following version, which gives an ambiguity error:

```python
from plum import dispatch

class Parent:
    pass

class B(Parent):
    pass

class BaseClass:
    pass
        
class Inheritor(BaseClass):
    pass

@dispatch
def foo(self: BaseClass, x: Parent):
    raise NotImplementedError

@dispatch
def foo(self: BaseClass, x: B):
    print("hello!")

@dispatch
def foo(self: Inheritor, x: Parent):
    print("cool!")
        
foo(BaseClass(), B()) 

foo(Inheritor(), B())
# AmbiguousLookupError: `foo(<__main__.Inheritor object at 0x7fbbc85661f0>, <__main__.B object at 0x7fbbc8566220>)` is ambiguous.
#
# Candidates:
#    foo(self: __main__.BaseClass, x: __main__.B)
#        <function foo at 0x7fbbc8548820> @ /<ipython-input-2-859b87ba2416>:19
#    foo(self: __main__.Inheritor, x: __main__.Parent)
#        <function foo at 0x7fbba82acdc0> @ /<ipython-input-2-859b87ba2416>:23
```

**Comment by femtomc** [comment]
Right, so the thing I want is that `Inheritor().foo(B())` prints `"hello!"` -- e.g. that switching the method on the parent class in the `Inheritor` would not affect the behavior of the specialization (that when I dispatch on `B`, I still get the original `"hello!"` behavior).

**Comment by femtomc** [comment]
The point being that, in `Inheritor` -- I could define _a different fallback on the parent class `Parent`_ while preserving specialized behavior which I inherit from `BaseClass`

**Comment by wesselb** [comment]
@femtomc, how would something like this look?

```python
from plum import dispatch

class Parent:
    pass

class B(Parent):
    pass

class BaseClass:
    pass
        
class Inheritor(BaseClass):
    pass

@dispatch
def foo_fallback(self: BaseClass, x: Parent):
    print("Fallback for BaseClass")

@dispatch
def foo_fallback(self: Inheritor, x: Parent):
    print("Fallback for Inheritor")

@dispatch
def foo(self: BaseClass, x: Parent):
    # If no specialisation is possible, run the fallback.
    return foo_fallback(self, x)

@dispatch
def foo(self: BaseClass, x: B):
    print("General specialisation")

        
foo(BaseClass(), Parent())  # Fallback for BaseClass
foo(BaseClass(), B())       # General specialisation

foo(Inheritor(), Parent())  # Fallback for Inheritor
foo(Inheritor(), B())       # General specialisation
```

**Comment by femtomc** [comment]
Interesting, in my classes -- I did not annotate `self`. Does this affect dispatch in this weird combination of OOP and MD?

**Comment by femtomc** [comment]
@PhilipVinc I wonder if it's useful to have _two_ dispatch patterns.

Multiple dispatch, as accessed via `dispatch` -- does the thing you'd expect, and behaves well with functions, etc.

But this sort of hybrid dispatch (maybe `class_dispatch`?) supports this hybrid style of OOP + MD? It has slightly different dispatch rules on the `self` argument.

**Comment by femtomc** [comment]
I actually wonder if I can just hack this up by defining a new abstract base class, which owns a `Dispatcher`...

**Comment by femtomc** [comment]
Okay -- consider the following two screenshots.

If I don't define a new method `foo` in the inheritor class, I do inherit the method (and all its implementations) from the base class:
![image](https://github.com/beartype/plum/assets/34410036/30d94c9c-f780-402e-86e7-43e56c8f2f31)

However, if I do define `foo` -- it just completely replaces the method.
![image](https://github.com/beartype/plum/assets/34410036/ff8bb6a1-ddfb-4818-a0f5-36191f2d24e1)

So, in a sense, `plum` is already taking a stance on inherited methods -- the rule is: if you don't define a method with the same name, you'll inherit all the methods from the base class, otherwise -- you clobber the method table and remove all the methods from the base class.

I think my argument is that a (possibly more) useful stance _is to merge_ the method tables. This is basically treating a base class like a "module" from Julia (for instance).

I'm aware that this will cause interesting scenarios: like, what happens if I inherit from two base classes that each define the same method? I think (just like in Julia), we should reject these scenarios as something which should cause an error.

**Comment by wesselb** [comment]
> So, in a sense, plum is already taking a stance on inherited methods -- the rule is: if you don't define a method with the same name, you'll inherit all the methods from the base class, otherwise -- you clobber the method table and remove all the methods from the base class.

Yes, you're right! When Plum was designed, we decided that we would handle inheritance in a way that would most naturally integrate with how Python handles inheritance: (1) Separate the method tables of class definitions. (2) When an appropriate method cannot be found, instead of raising a `NotFoundLookupError`, attempt the function of the parent according to `Class.mro()`.

I agree that merging the method tables would be very interesting and could enable interesting new design patterns. It would, however, deviate from how method lookup in classes currently works (_i.e._ according to the MRO), which could be confusing. To decide whether this would be worth it, I guess one should figure out which new design patterns would be possible and whether those really aren't possible with the current model.
